plutus-tx/src/PlutusTx/Builtins.hs

```diff
@@ -385,7 +385,7 @@ encodeUtf8 = BI.encodeUtf8
 
 {-# INLINABLE matchList #-}
 matchList :: forall a r . BI.BuiltinList a -> r -> (a -> BI.BuiltinList a -> r) -> r
-matchList l nilCase consCase = BI.chooseList l (const nilCase) (\_ -> consCase (BI.head l) (BI.tail l)) ()
+matchList l ~nilCase ~consCase = BI.chooseList l (const nilCase) (\_ -> consCase (BI.head l) (BI.tail l)) ()
```

Comment:
Here we explicitly rely on GHC inlining this definition for us. Can we instead make sure to inline it ourselves? Is there a general rule? How many other such cases are we missing?

It sucks that we don't want GHC to touch our code much (because it doesn't understand the semantics), while at the time we want to rely on it in some cases.